androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LocationProviderGmsCore.java

```diff
@@ -0,0 +1,190 @@
+
+    // Values for the LocationRequest's setInterval for normal and high accuracy, respectively.
+    private static final long UPDATE_INTERVAL_MS = 1000;
+    private static final long UPDATE_INTERVAL_FAST_MS = 500;
+
+    private static LocationProviderGmsCore sProvider;
+    private final GoogleApiClient mGoogleApiClient;
+    private FusedLocationProviderApi mLocationProviderApi = LocationServices.FusedLocationApi;
+
+    private TrustedWebActivityCallbackRemote mCallback;
+
+    private boolean mEnablehighAccuracy;
+    private LocationRequest mLocationRequest;
+
+    LocationProviderGmsCore(Context context) {
+        mGoogleApiClient = new GoogleApiClient.Builder(context)
+                .addApi(LocationServices.API)
+                .addConnectionCallbacks(this)
+                .addOnConnectionFailedListener(this)
+                .build();
+        assert mGoogleApiClient != null;
+    }
+
+    public static boolean isGooglePlayServicesAvailable(Context context) {
+        return GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context)
+                == ConnectionResult.SUCCESS;
+    }
+
+    public static LocationProviderGmsCore getInstance(Context context) {
+        if (sProvider == null) {
+            sProvider = new LocationProviderGmsCore(context);
+        }
+        return sProvider;
+    }
+
+    // ConnectionCallbacks implementation
+    @Override
+    public void onConnected(Bundle connectionHint) {
+        mLocationRequest = LocationRequest.create();
+        if (mEnablehighAccuracy) {
+            // With enableHighAccuracy, request a faster update interval and configure the provider
+            // for high accuracy mode.
+            mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)
+                    .setInterval(UPDATE_INTERVAL_FAST_MS);
+        } else {
+            mLocationRequest.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY)
+                    .setInterval(UPDATE_INTERVAL_MS);
+        }
+
+        final Location location = mLocationProviderApi.getLastLocation(mGoogleApiClient);
+        if (location != null) {
+            onNewLocationAvailable(location);
+        }
+
+        try {
+            // Request updates on UI Thread replicating LocationProviderAndroid's behaviour.
+            mLocationProviderApi.requestLocationUpdates(
+                    mGoogleApiClient, mLocationRequest, this, Looper.getMainLooper());
+        } catch (IllegalStateException | SecurityException e) {
+            // IllegalStateException is thrown "If this method is executed in a thread that has not
+            // called Looper.prepare()". SecurityException is thrown if there is no permission, see
+            // https://crbug.com/731271.
+            Log.e(TAG, " mLocationProviderApi.requestLocationUpdates() " + e);
+            notifyLocationErrorWithMessage(
+                    "Failed to request location updates: " + e.toString());
+            assert false;
```

Comment:
Do we really want to crash here?

(Also, while asserts in Java are turned on in Chrome, I think in general they are not, so asserts in the Android Browser Helper will have no effect.)